web/ui/src/common/resourcesFactory.js

```diff
@@ -269,16 +287,56 @@
 
             return function(/* args */){
                 var url = config.url.apply(null, arguments),
-                    payload;
+                    resourceName = url,
+                    payload,
+                    // deferred that will be returned to the user
+                    deferred = $q.defer();
+
+                // if resourceName has query params, strip em off
+                if(resourceName.indexOf("?")){
+                    resourceName = resourceName.split("?")[0];
+                }
+
+                // NOTE: all of our code expects angular's wrapped
+                // promises which provide a success and error method
+                // TODO - remove the need for this
+                httpify(deferred);
+
+                if(pendingGETRequests[resourceName]){
+                    // theres already a pending request to
+                    // this endpoint, so fail!
+                    deferred.reject(`a request to ${resourceName} is pending`);
+                    return deferred.promise;
+                }
 
                 if(config.payload){
                     payload = config.payload.apply(null, arguments);
                 }
 
-                return $http[config.method](url, payload)
-                    .error(function(data, status) {
-                        redirectIfUnauthorized(status);
-                    });
+                $http({
+                    method: config.method,
+                    url: url,
+                    data: payload,
+                    timeout: REQUEST_TIMEOUT
+                })
+                .success(function(data, status){
+                    deferred.resolve(data);
+                })
+                .error(function(data, status) {
+                    // TODO - include data as well?
+                    deferred.reject(status);
+                    redirectIfUnauthorized(status);
+                })
+                .finally(function() {
+                    pendingGETRequests[resourceName] = null;
```

Comment:
Ditto the previous comment about checking the method. We don't want a POST or DELETE to remove a cached promise for a GET that's still pending